spec/routes/api/vm_spec.rb

```diff
@@ -101,17 +288,136 @@
           unix_user: "ubi",
           size: "standard-2",
           boot_image: "ubuntu-jammy"
-        }
+        }.to_json
 
         expect(last_response.status).to eq(400)
         expect(JSON.parse(last_response.body)["error"]["details"]["billing_info"]).to eq("Project doesn't have valid billing information")
       end
+
+      it "invalid body" do
+        post "/api/project/#{project.ubid}/location/#{TEST_LOCATION}/vm/test-vm", "invalid_body"
+
+        expect(last_response.status).to eq(400)
+        expect(JSON.parse(last_response.body)["error"]["details"]["body"]).to eq("Body isn't a valid JSON object.")
+      end
+
+      it "missing required key" do
+        post "/api/project/#{project.ubid}/location/#{TEST_LOCATION}/vm/test-vm", {
+          unix_user: "ubi"
+        }.to_json
+
+        expect(last_response.status).to eq(400)
+        expect(JSON.parse(last_response.body)["error"]["details"]["parameters"]).to eq("Request body must include required parameters: public_key")
+      end
+
+      it "non allowed key" do
+        post "/api/project/#{project.ubid}/location/#{TEST_LOCATION}/vm/test-vm", {
+          public_key: "ssh key",
+          foo_key: "foo_val"
+        }.to_json
+
+        expect(last_response.status).to eq(400)
+        expect(JSON.parse(last_response.body)["error"]["details"]["parameters"]).to eq("Only following parameters are allowed: public_key, size, unix_user, boot_image, enable_ip4, private_subnet_id")
+      end
+
+      it "firewall-rule" do
+        post "/api/project/#{project.ubid}/location/#{vm.location}/vm/#{vm.name}/firewall-rule", {
+          cidr: "0.0.0.0/0",
+          port_range: "100..101"
+        }.to_json
+
+        expect(last_response.status).to eq(200)
+      end
+
+      it "firewall-rule vm ubid" do
+        post "/api/project/#{project.ubid}/location/#{vm.location}/vm/ubid/#{vm.ubid}/firewall-rule", {
+          cidr: "0.0.0.0/0",
+          port_range: "100..1012"
+        }.to_json
+
+        expect(last_response.status).to eq(200)
+      end
+
+      it "firewall-rule no port range" do
+        post "/api/project/#{project.ubid}/location/#{vm.location}/vm/#{vm.name}/firewall-rule", {
+          cidr: "0.0.0.0/1"
+        }.to_json
+
+        expect(last_response.status).to eq(200)
+      end
+
+      it "firewall-rule single port" do
+        post "/api/project/#{project.ubid}/location/#{vm.location}/vm/#{vm.name}/firewall-rule", {
+          cidr: "0.0.0.0/1",
+          port_range: "11111"
+        }.to_json
+
+        expect(last_response.status).to eq(200)
+      end
+
+      it "firewall-rule invalid cidr" do
+        post "/api/project/#{project.ubid}/location/#{vm.location}/vm/#{vm.name}/firewall-rule", {
+          cidr: "0.0.0"
+        }.to_json
+
+        expect(last_response.status).to eq(400)
+        expect(JSON.parse(last_response.body)["error"]["details"]["cidr"]).to eq("Invalid cidr")
+      end
+
+      it "firewall-rule invalid port range smaller single" do
```

Comment:
IMHO, these tests shouldn't be part of the VM spec. We should thoroughly test validation rules in the validation module. The branches belong to `Validation.validate_port_range`, not the VM listing route. However, I'm not certain if my perspective is correct for proper unit testing. @fdr What do you think about it? It's more about our testing philosophy.